solver/src/solver/single_order_solver.rs

```diff
@@ -59,40 +53,31 @@ impl<I: SingleOrderSolving + Send + Sync + 'static> Solver for SingleOrderSolver
                 Liquidity::Limit(order) => Some(order),
                 _ => None,
             })
-            .collect::<Vec<_>>();
-
-        // Randomize which orders we take, this prevents this solver "getting
-        // stuck" on bad orders.
-        if orders.len() > MAX_SETTLEMENTS {
-            orders.shuffle(&mut rand::thread_rng());
-        }
-
-        let settlements = future::join_all(
-            orders
-                .into_iter()
-                .take(MAX_SETTLEMENTS)
-                .map(|order| tokio::time::timeout_at(deadline.into(), self.settle_order(order))),
-        )
-        .await;
-
-        Ok(settlements
-            .into_iter()
-            .filter_map(|settlement| match settlement {
-                Err(_deadline) => {
-                    tracing::error!("inner solver {:?} timeout", self.name());
-                    None
-                }
-                Ok(Ok(Some(settlement))) => Some(settlement),
-                Ok(Ok(None)) => None,
-                Ok(Err(err)) => {
-                    // It could be that the inner solver can't match an order and would
-                    // return an error for whatever reason. In that case, we want
-                    // to continue trying to solve for other orders.
-                    tracing::error!("{} inner solver error: {:?}", self.name(), err);
-                    None
+            .collect::<VecDeque<_>>();
+        // Randomize which orders we start with to prevent us getting stuck on bad orders.
+        orders.make_contiguous().shuffle(&mut rand::thread_rng());
+
+        let mut settlements = Vec::new();
+        let settle = async {
+            while let Some(order) = orders.pop_front() {
+                match self.inner.try_settle_order(order.clone()).await {
+                    Ok(settlement) => settlements.extend(settlement),
+                    Err(err) => {
+                        let name = self.inner.name();
+                        if err.retryable {
+                            tracing::warn!("Solver {} benign error: {:?}", name, &err);
+                            orders.push_back(order);
+                        } else {
+                            tracing::error!("Solver {} hard error: {:?}", name, &err);
```

Comment:
Maybe we can increment some metrics instead of alerting with error here. Ok with doing that in a separate PR.